packages/zoe/src/contractFacet/contractFacet.js

```diff
@@ -0,0 +1,348 @@
+import { E } from '@agoric/eventual-send';
+import makeWeakStore from '@agoric/weak-store';
+
+import makeAmountMath from '@agoric/ertp/src/amountMath';
+import { makeNotifierKit, updateFromNotifier } from '@agoric/notifier';
+import { makePromiseKit } from '@agoric/promise-kit';
+import { areRightsConserved } from './rightsConservation';
+import { makeIssuerTable } from '../issuerTable';
+import {
+  assertKeywordName,
+  getKeywords,
+  cleanProposal,
+} from '../zoeService/cleanProposal';
+import { evalContractBundle } from './evalContractCode';
+import { makeZcfSeatAdminKit } from './seat';
+import { makeExitObj } from './exit';
+import { objectMap } from '../objArrayConversion';
+
+import '../../exported';
+import '../internal-types';
+
+export function buildRootObject() {
+  /** @type ExecuteContract */
+  const executeContract = async (
+    bundle,
+    zoeService,
+    invitationIssuer,
+    zoeInstanceAdmin,
+    instanceRecord,
+  ) => {
+    const issuerTable = makeIssuerTable();
+    const getAmountMath = brand => issuerTable.get(brand).amountMath;
+
+    const invitationHandleToHandler = makeWeakStore('invitationHandle');
+
+    /** @type WeakStore<ZCFSeat,ZCFSeatAdmin> */
+    const seatToZCFSeatAdmin = makeWeakStore('seat');
+
+    const issuers = Object.values(instanceRecord.issuerKeywordRecord);
+
+    const getPromiseForIssuerRecords = issuersP =>
+      Promise.all(issuersP.map(issuerTable.getPromiseForIssuerRecord));
+
+    await getPromiseForIssuerRecords(issuers);
+
+    const allSeatStagings = new WeakSet();
+
+    const registerIssuerRecord = (keyword, issuerRecord) => {
+      assertKeywordName(keyword);
+      assert(
+        !getKeywords(instanceRecord.issuerKeywordRecord).includes(keyword),
+        details`keyword ${keyword} must be unique`,
+      );
+      instanceRecord = {
+        ...instanceRecord,
+        issuerKeywordRecord: {
+          ...instanceRecord.issuerKeywordRecord,
+          [keyword]: issuerRecord.issuer,
+        },
+        brandKeywordRecord: {
+          ...instanceRecord.brandKeywordRecord,
+          [keyword]: issuerRecord.brand,
+        },
+      };
+
+      return issuerRecord;
+    };
+
+    /** @type MakeZCFMint */
+    const makeZCFMint = async (keyword, mathHelperName = 'nat') => {
+      assert(
+        !(keyword in instanceRecord.issuerKeywordRecord),
+        details`Keyword ${keyword} already registered`,
+      );
+
+      const zoeMintP = E(zoeInstanceAdmin).makeZoeMint(keyword, mathHelperName);
+      const { brand: mintyBrand, issuer: mintyIssuer } = await E(
+        zoeMintP,
+      ).getIssuerRecord();
+      // AWAIT
+      const mintyAmountMath = makeAmountMath(mintyBrand, mathHelperName);
+      const mintyIssuerRecord = harden({
+        brand: mintyBrand,
+        issuer: mintyIssuer,
+        amountMath: mintyAmountMath,
+      });
+      registerIssuerRecord(keyword, mintyIssuerRecord);
+      issuerTable.registerIssuerRecord(mintyIssuerRecord);
```

Comment:
I should not have given these two functions the same name. Although elsewhere we have a registerIssuerByKeyword, we have no conflict with registerIssuerRecordByKeyword. But please use a shorter name if possible.

```suggestion
      const mintyAmountMath = makeAmountMath(mintyBrand, mathHelperName);
      const mintyIssuerRecord = harden({
        brand: mintyBrand,
        issuer: mintyIssuer,
        amountMath: mintyAmountMath,
      });
      registerIssuerRecordByKeyword(keyword, mintyIssuerRecord);
      issuerTable.registerIssuerRecord(mintyIssuerRecord);
```

I notice that lines 91-96,98 are equivalent to issuerTable.js lines 81-87. Perhaps more of this common logic should move into `issuerTable.registerIssuerRecord` (which might then have a different name anyway)?

Reply:
Renamed. I put off the discussion of the logic being moved to issuerTable, since that may be changing drastically anyway